File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        i,j = 0,1
        n = len(prices)
        res = 0
        while (i<j and j<n):
            profit=prices[j]-prices[i]
            if profit < 0:
                i=j
                j+=1
                res = max(res,profit)
            else:
                res = max(res,profit)
                j+=1
        return res
        
        
        

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int ans = 0;
        int l = 0, r = l+1;
        while(l < r && r < prices.length){
            int buy = prices[l];
            int sell = prices[r];
            int profit = sell - buy;
            ans = Math.max(ans, profit);
            r++;
            // System.out.println(String.format("buy %d, sell %d, ans = %d",buy,sell,ans));
            if(r == prices.length){
                while(l < r && prices[l] >= buy){
                    l++;
                    // System.out.println("\tincrementing l="+l);
                }
                // System.out.println("\tbuy "+prices[l]);
                r = l+1;
            }
        }
        return ans;
    }
}